tests/test_formatter.py

```diff
@@ -36,3 +40,62 @@ def test_problem_formatter(problem_parser, pddl_file):
     actual_problem_str = problem_to_string(expected_problem_obj)
     actual_problem_obj = problem_parser(actual_problem_str)
     assert actual_problem_obj == expected_problem_obj
+
+
+def test_typed_constants_formatting_in_domain() -> None:
+    """Test that types and typed constants are formatted correctly."""
+    t1, t2, t3 = "type_1", "type_2", "type_3"
+
+    a1, b1, c1 = constants("a b c", type_=t1)
+    d2, e2, f2 = constants("d e f", type_=t2)
+    g3, h3, i3 = constants("g h i", type_=t3)
+    j, k, lc = constants("j k l", type_=None)
+
+    # define the domain object.
+    domain = Domain(
+        "my_domain",
+        requirements=[Requirements.TYPING],
+        types={t1: None, t2: t1, t3: t1},
+        constants=[a1, b1, c1, d2, e2, f2, g3, h3, i3, j, k, lc],
+    )
+
+    domain_str = domain_to_string(domain)
+
+    assert domain_str == dedent(
+        """\
+    (define (domain my_domain)
+        (:requirements :typing)
+        (:types type_2 type_3 - type_1 type_1)
```

Comment:
Note, earlier we did not consider types with no parents, and we printed `type_1` in the first position. However, this is not correct wrt PDDL syntax; typed names with no type should be at the end of the list, o/w there is no way the parsing rules can spot them:
![image](https://github.com/AI-Planning/pddl/assets/9467617/30ec8961-5f75-475a-9d8d-ea0a8f3d1aaa)